libmount/src/hook_loopdev.c

```diff
@@ -518,6 +532,47 @@ static int hook_prepare_loopdev(
 	return rc;
 }
 
+static int hookset_mkerrmsg(
+                        struct libmnt_context *cxt,
+                        const struct libmnt_hookset *hs,
+                        char *buf, size_t sz)
+{
+	int er = cxt->failure_errno;
+
+	DBG(HOOK, ul_debugobj(hs, "calling mkerrmsg"));
+
+	/* parse errors */
+	if (er == EINVAL) {
+		switch (cxt->failure_cmd) {
+		case MNT_MS_OFFSET:
+		case MNT_MS_SIZELIMIT:
```

Comment:
This mechanism disconnects the location where the error is encountered in the code from the message that is generated.
So to go from an error message to the respective code location developers have to work backwards from this switch-case statement.

What about formatting the message as part of the call to `mnt_context_save_failure()`?

This would also remove the need to figure unique error discriminants for each error site.